### ISSUE TITLE
Does  geNomad currently support training new models?

### ISSUE THREAD

**Issue by a-piece-of-teemo** [apcamargo/genomad#52]

**Comment by apcamargo** [comment]
There's no command that provides an interface for quickly training a model. But if you feel adventurous, you could definitely train new models. All the data necessary to train models is being saved (see the features files generated by `marker-classification` and the encoded sequences generated by `nn-classification`).

**Comment by a-piece-of-teemo** [comment]
Excuse me,would it be possible for you to provide the script used for training this model for me?

**Comment by apcamargo** [comment]
Sure!

You can get the code to train the neural network model here: https://zenodo.org/records/8049246

I don't have a structured script to train the XGboost model used in the marker branch. But it should be very easy to do. All the features are stored in the `<prefix>_marker_classification/<prefix>_features.npz` file (you can see how to read it [here](https://github.com/apcamargo/genomad/blob/b90fb3b24f176f0fdf1719cc848b57cede7fc95f/genomad/modules/marker_classification.py#L482-L492)). You could even add additional features to the model, such as the number of plasmid/virus hallmarks (which are computed, but not used by the model to classify the sequences).

Here are the hyperparameters I to create the model:
```python
model = xgb.XGBClassifier(use_label_encoder=False, eval_metric="mlogloss", eta=0.2, max_depth=10, n_estimators=135)
```

Out of curiosity, is there a specific reason you want to train new models? If that's something with a broad appeal, I can consider adding in a future update.

**Comment by a-piece-of-teemo** [comment]
Thank you for your response and help！
I want to train a new model without any specific purpose, I just want to learn some relevant knowledge.

**Comment by Dx-wmc** [comment]
> 当然！
> 
> 您可以在此处获取训练神经网络模型的代码：[https ://zenodo.org/records/8049246](https://zenodo.org/records/8049246)
> 
> 我没有结构化脚本来训练标记分支中使用的 XGboost 模型。但这应该很容易做到。所有功能都存储在文件中（您可以[在此处](https://github.com/apcamargo/genomad/blob/b90fb3b24f176f0fdf1719cc848b57cede7fc95f/genomad/modules/marker_classification.py#L482-L492)`<prefix>_marker_classification/<prefix>_features.npz`查看如何读取它）。您甚至可以向模型添加其他功能，例如质粒/病毒标志的数量（这些标志是计算出来的，但模型不使用它来对序列进行分类）。[](https://github.com/apcamargo/genomad/blob/b90fb3b24f176f0fdf1719cc848b57cede7fc95f/genomad/modules/marker_classification.py#L482-L492)
> 
> 以下是我创建模型的超参数：
> 
> ```python
> model = xgb.XGBClassifier(use_label_encoder=False, eval_metric="mlogloss", eta=0.2, max_depth=10, n_estimators=135)
> ```
> 
> 出于好奇，您想训练新模型有什么具体原因吗？如果这具有广泛的吸引力，我可以考虑在未来的更新中添加。

Is it possible to add transposons and insertion sequences in the future? Predicting these MGEs based on machine learning may have better performance than current tools.

**Comment by Dx-wmc** [comment]
> Sure!
> 
> You can get the code to train the neural network model here: https://zenodo.org/records/8049246
> 
> I don't have a structured script to train the XGboost model used in the marker branch. But it should be very easy to do. All the features are stored in the `<prefix>_marker_classification/<prefix>_features.npz` file (you can see how to read it [here](https://github.com/apcamargo/genomad/blob/b90fb3b24f176f0fdf1719cc848b57cede7fc95f/genomad/modules/marker_classification.py#L482-L492)). You could even add additional features to the model, such as the number of plasmid/virus hallmarks (which are computed, but not used by the model to classify the sequences).
> 
> Here are the hyperparameters I to create the model:
> 
> ```python
> model = xgb.XGBClassifier(use_label_encoder=False, eval_metric="mlogloss", eta=0.2, max_depth=10, n_estimators=135)
> ```
> 
> Out of curiosity, is there a specific reason you want to train new models? If that's something with a broad appeal, I can consider adding in a future update.

Is it possible to add transposons and insertion sequences in the future? Predicting these MGEs based on machine learning may have better performance than current tools.

**Comment by ZoeLct** [comment]
Dear author, I apologize for disturbing you, but I would like to consult you about some issues I am facing with training this xgboost model:
I aim to train a model with performance similar to yours, so I used the train_test_sequences.fna from your benchmark_data directory. Then, using your script, I generated the train_test_sequences_marker_classification.tsv and npz files, with the contents as follows. Based on my understanding, I should select the features 'strand_switch_rate', 'coding_density', 'no_rbs_freq', 'sd_bacteroidetes_rbs_freq', 'sd_canonical_rbs_freq', 'tatata_rbs_freq', 'cc_marker_freq', 'cp_marker_freq', 'cv_marker_freq', 'pc_marker_freq', 'pp_marker_freq', 'pv_marker_freq', 'vc_marker_freq', 'vp_marker_freq', 'vv_marker_freq', 'c_marker_freq', 'p_marker_freq', 'v_marker_freq', 'median_c_spm', 'median_p_spm', 'median_v_spm', 'v_vs_c_score_logistic', 'v_vs_p_score_logistic', 'p_vs_c_score_logistic', 'gv_marker_freq', and 'seq_name' as the samples for training the xgboost model.
However, I have some questions and would appreciate your answers:
1、The steps mentioned above were completed based on my experience; if there are any issues, please let me know.
2、According to your annotation, train_test_sequences.fna.gz is used for training geNomad and for benchmarking its performance, consisting of nucleotide sequences of genome fragments produced by slicing genome sequences into metagenome-sized fragments. However, I am unclear about the purpose of the sequence_weights.tsv file, which, according to its annotation, is used to assign weights to fragments during training. Typically, I have only seen label sets without weights, so I am unsure how to use this file. Please tell me how to use it.

**Comment by ZoeLct** [comment]
3、The following code is what I wrote based on the paper and my understanding for training an xgboost model. Since I have never worked with the xgboost model before, there are many errors in it, but I am also unclear on how to correct them. I hope you can help me identify the errors in the code and my misunderstandings. If possible, I would appreciate it if you could provide me with a script that can run correctly for training xgboost.
import pandas as pd
import xgboost as xgb
from shaphypetune import BoostSearch, BoostBoruta
from joblib import dump

# Load feature data
features_df = pd.read_csv('train_test_sequences_features.tsv', sep='\t', usecols=[
    'seq_name', 'strand_switch_rate', 'coding_density', 'no_rbs_freq', 
    'sd_bacteroidetes_rbs_freq', 'sd_canonical_rbs_freq', 'tatata_rbs_freq', 
    'cc_marker_freq', 'cp_marker_freq', 'cv_marker_freq', 'pc_marker_freq', 
    'pp_marker_freq', 'pv_marker_freq', 'vc_marker_freq', 'vp_marker_freq', 
    'vv_marker_freq', 'c_marker_freq', 'p_marker_freq', 'v_marker_freq', 
    'median_c_spm', 'median_p_spm', 'median_v_spm', 'v_vs_c_score_logistic', 
    'v_vs_p_score_logistic', 'p_vs_c_score_logistic', 'gv_marker_freq'
])

# Load weights and labels data
weights_labels_df = pd.read_csv('sequence_weights.tsv', sep='\t', usecols=[
    'seq_name', 'label', 'Weight in split 1', 'Weight in split 2', 
    'Weight in split 3', 'Weight in split 4', 'Weight in split 5'
])

# Merge datasets
data_df = pd.merge(features_df, weights_labels_df, on='seq_name')
# Separate features and labels
X = data_df.drop(['seq_name', 'label'], axis=1)
y = data_df['label']

# Separate weights (using split 1 as an example)
weights = data_df['Weight in split 1']
# Initialize XGBClassifier model
model = xgb.XGBClassifier(use_label_encoder=False, eval_metric="mlogloss", eta=0.2, max_depth=10, n_estimators=135)

# Perform feature selection with BoostBoruta
boruta_selector = BoostBoruta(
    estimator=model,
    perc=100,
    alpha=0.05,
    max_iter=100,
    importance_type='shap',
    train_importance=True,
    verbose=2,
    n_jobs=-1
)

# Identify important features
X_filtered = boruta_selector.fit_transform(X, y)

# Set up the parameter grid
param_grid = {
    'eta': [0.01, 0.1, 0.2],
    'max_depth': [6, 10, 15],
    'n_estimators': [100, 135, 150]
}

# Initialize BoostSearch
boost_search = BoostSearch(
    estimator=model,
    param_grid=param_grid,
    greater_is_better=False,
    n_iter=10,
    sampling_seed=42,
    verbose=2,
    n_jobs=-1
)

# Execute grid search
best_model = boost_search.fit(X_filtered, y, sample_weight=weights)

# Train the model using the best parameters
final_model = best_model.best_estimator_
final_model.fit(X_filtered, y, sample_weight=weights)

# Save the model
dump(final_model, 'xgboost_model.ubj')

I apologize for the inconvenience, but I have encountered many problems in training this xgboost model, and I hope you can answer my questions. Some issues that may seem simple to you are quite challenging for me to resolve, and your assistance would be greatly appreciated.

**Comment by apcamargo** [comment]
The procedure you described makes sense. I think you are in the right direction. Just note that `seq_name` shouldn't be a feature. It's just the name of the sequence and it's not used by the model.

> According to your annotation, train_test_sequences.fna.gz is used for training geNomad and for benchmarking its performance, consisting of nucleotide sequences of genome fragments produced by slicing genome sequences into metagenome-sized fragments. However, I am unclear about the purpose of the sequence_weights.tsv file, which, according to its annotation, is used to assign weights to fragments during training. Typically, I have only seen label sets without weights, so I am unsure how to use this file. Please tell me how to use it.

Weighting classes is indeed more common, but it is possible to assign different weights for each sample (`sample_weight` parameter in XGBoost). I weighted samples differently to account for taxonomic biases in the dataset (you can read more about this procedure in geNomad's paper).

> The following code is what I wrote based on the paper and my understanding for training an xgboost model. Since I have never worked with the xgboost model before, there are many errors in it, but I am also unclear on how to correct them. I hope you can help me identify the errors in the code and my misunderstandings. If possible, I would appreciate it if you could provide me with a script that can run correctly for training xgboost.

I couldn't find any obvious mistakes in the code. What errors are you getting?

**Comment by ZoeLct** [comment]
Dear author, I have some questions to ask you:
1、Regarding the weight in split1-5 in your sequence_weights.tsv file, does it correspond to train_test_sequences.fna.gz as well as the generated features.tsv and npz, meaning that the entire dataset and features are divided into 5 different groups of weights, with each group of weights being independent of each other? When training the code, is it that only one group of weights is used at a time, for example, selecting only Weight in split 0 weights = data_df['Weight in split 0']?
2、In the code I wrote above, is the use of the weights part of the code correct? I have not previously worked with model training that involves weights, and I would like to know if there are any errors in my code or anything that needs to be modified?
3、If you also notice any minor issues in the code, please let me know. I have just started working with the xgboost model, and there are many small details that I do not quite understand.
Sorry to bother you, and I hope to receive your reply.
